book-content/chapters/04-essential-types-and-annotations.md

```diff
@@ -0,0 +1,2366 @@
+
+```ts
+const logAlbumInfo = (
+  title: string,
+  trackCount: number,
+  isReleased: boolean,
+  releaseDate?: string,
+) => {
+  // rest of function body
+};
+```
+
+Now we can call `logAlbumInfo` and include a release date string, or leave it out:
+
+```ts
+logAlbumInfo("Midnights", 13, true, "2022-10-21");
+
+logAlbumInfo("American Beauty", 10, true);
+```
+
+Hovering over the optional `releaseDate` parameter in VS Code will show us that it is now typed as `string | undefined`.
+
+We'll discuss the `|` symbol more later, but this means that the parameter could either be a `string` or `undefined`. It would be acceptable to literally pass `undefined` as a second argument, or it can be omitted all together.
+
+### Default Parameters
+
+In addition to marking parameters as optional, you can set default values for parameters by using the `=` operator.
+
+For example, we could set the `format` to default to `"CD"` if no format is provided:
+
+```ts
+const logAlbumInfo = (
+  title: string,
+  trackCount: number,
+  isReleased: boolean,
+  format: string = "CD",
+) => {
+  // rest of function body
+};
+```
+
+The annotation of `: string` can also be omitted:
+
+```ts
+const logAlbumInfo = (
+  title: string,
+  trackCount: number,
+  isReleased: boolean,
+  format = "CD",
+) => {
+  // rest of function body
+};
+```
+
+Since it can infer the type of the `format` parameter from the value provided. This is another nice example of type inference.
+
+### Function Return Types
+
+In addition to setting parameter types, we can also set the return type of a function.
+
+The return type of a function can be annotated by placing a `:` and the type after the closing parentheses of the parameter list. For the `logAlbumInfo` function, we can specify that the function will return a string:
+
+```ts
+const logAlbumInfo = (
+  title: string,
+  trackCount: number,
+  isReleased: boolean,
+): string => {
+  // rest of function body
+};
+```
+
+If the value returned from a function doesn't match the type that was specified, TypeScript will show an error.
+
+```ts twoslash
+// @errors: 2322
+const logAlbumInfo = (
+  title: string,
+  trackCount: number,
+  isReleased: boolean,
+): string => {
+  return 123;
+};
+```
+
+Return types are useful for when you want to ensure that a function returns a specific type of value.
+
+### Rest Parameters
+
+Just like in JavaScript, TypeScript supports rest parameters by using the `...` syntax for the final parameter. This allows you to pass in any number of arguments to a function.
+
+For example, this `printAlbumFormats` is set up to accept an `album` and any number of `formats`:
+
+```ts
+function getAlbumFormats(album: Album, ...formats: string[]) {
+  return `${album.title} is available in the following formats: ${formats.join(
+    ", ",
+  )}`;
+}
+```
+
+Declaring the parameter with the `...formats` syntax combined with an array of strings lets us pass in any number of strings to the function:
```

Comment:
might be worth mentioning somewhere nearby that "any" here means between 0 and infinity